File: src/main/java/sage/web/UserController.java
package sage.web;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.multipart.MultipartFile;
import sage.service.UserService;
import sage.transfer.UserCard;
import sage.transfer.UserSelf;
import sage.web.auth.Auth;

@RestController
@RequestMapping("/user")
public class UserController {
  @Autowired
  UserService userService;

  @RequestMapping("/self")
  public UserSelf self() {
    Long uid = Auth.checkCuid();
    return userService.getSelf(uid);
  }

  @RequestMapping("/card/{id}")
  public UserCard userCard(@PathVariable Long id) {
    Long uid = Auth.checkCuid();
    return userService.getUserCard(uid, id);
  }

  @RequestMapping("/info/{id}")
  Object userInfo(@PathVariable Long id) {
    throw new UnsupportedOperationException();
  }

  @RequestMapping(value = "/change-intro", method = RequestMethod.POST)
  void changeIntro(@RequestParam String intro) {
    userService.changeIntro(Auth.checkCuid(), intro);
  }

  @RequestMapping(value = "/change-avatar", method = RequestMethod.POST)
  void changeAvatar(@RequestParam MultipartFile photo) {
    userService.changeAvatar(Auth.checkCuid(), null);
  }

}

File: src/main/kotlin/sage/web/page/UserController.kt
package sage.web.page

import org.springframework.stereotype.Controller
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.servlet.ModelAndView
import sage.entity.Blog
import sage.entity.User
import sage.transfer.BlogPreview
import sage.web.auth.Auth
import sage.web.context.BaseController

@Controller
@RequestMapping("/users")
class UserController : BaseController() {

  @RequestMapping("/self")
  fun userPage() = "redirect:/users/${Auth.checkUid()}"

  @RequestMapping("/{id}")
  fun userPage(@PathVariable id: Long): ModelAndView {
    val uid = Auth.uid()
    val thisUser = userService.getUserCard(uid, id)
    val blogs = Blog.byAuthor(id).map(::BlogPreview)

    frontMap().attr("id", id).attr("isSelfPage", uid == id)

    return ModelAndView("user-page").addObject("thisUser", thisUser)
        .addObject("blogs", blogs)
  }

  @RequestMapping("/{id}/card")
  fun userCard(@PathVariable id: Long): ModelAndView {
    val uid = Auth.checkUid()
    val theUser = userService.getUserCard(uid, id)
    return ModelAndView("user-card").addObject("user", theUser)
  }

  @RequestMapping
  fun people(): ModelAndView {
    val uid = Auth.uid()
    val recomms = if (uid != null) userService.recommendByTag(uid) else emptyList()
    val people = userService.people(uid)
    return ModelAndView("people").addObject("recomms", recomms).addObject("people", people)
  }

  @RequestMapping("/{id}/rss")
  fun rss(@PathVariable id: Long): ModelAndView {
    val blogs = blogService.authorRSS(id)
    response.contentType = "text/xml"
    return ModelAndView("rss").addObject("blogs", blogs).addObject("name", User.get(id).name)
  }
}
